File: array/3_kth_min_max.c
#include <stdio.h>

void swap(int *arr, int x, int y)
{
    int temp = arr[x];
    arr[x] = arr[y];
    arr[y] = temp;
}

int partition(int *arr, int start, int end)
{
    int pivot = arr[end];
    int tail_index = start;

    for (int i = start; i < end; i++)
    {
        if (arr[i] < pivot)
        {
            swap(arr, i, tail_index);
            tail_index++;
        }
    }
    swap(arr, end, tail_index);
    return tail_index;
}

int find_kth_max(int k, int *arr, int size)
{
    if ((k < 1) || (k > size))
    {
        printf("\nInvalid Input!");
        return -1;
    }
    else
    {
        int start = 0;
        int end = size - 1;

        while (start <= end)
        {
            int pivod_index = partition(arr, start, end);

            if (pivod_index == (size - k))
                return arr[pivod_index];

            if (pivod_index < (size - k))
                start = pivod_index + 1;
            else
                end = pivod_index - 1;
        }
    }
}

int find_kth_min(int k, int *arr, int size)
{
    if ((k < 1) || (k > size))
    {
        printf("\nInvalid Input!");
        return -1;
    }
    else
    {
        int start = 0;
        int end = size - 1;

        while (start <= end)
        {
            int pivod_index = partition(arr, start, end);

            if (pivod_index == (k - 1))
                return arr[pivod_index];

            if (pivod_index < (k - 1))
                start = pivod_index + 1;
            else
                end = pivod_index - 1;
        }
    }
}

void main()
{
    int arr[] = {10, 2, 7, 4, 9, 6, 3, 8, 1, 5};
    int size = sizeof(arr) / sizeof(int);
    int k;
    printf("\n ENTER K : ");
    scanf("%d", &k);
    printf("\n K-th MIN : %d ", find_kth_min(k, arr, size));
    printf("\n K-th MAX : %d ", find_kth_max(k, arr, size));
    printf("\n");
}

File: python3/array/3_kth_min_max.py
def partition(arr,start,end):
    pivot=arr[end]
    tail_index=start

    for i in range(start,end):
        if arr[i]<pivot:
            arr[i],arr[tail_index]=arr[tail_index],arr[i]
            tail_index+=1
        
    arr[end],arr[tail_index]=arr[tail_index],arr[end]
    return tail_index

def find_kth_min(k,arr):
    if k<1 or k>len(arr):
        print ('Invalid Input!')
        return -1
    else:
        start=0
        end=len(arr)-1
        while start<=end:
            pivot_index=partition(arr,start,end)
            if pivot_index==(k-1):
                return arr[pivot_index]
            elif pivot_index<(k-1):
                start=pivot_index+1
            else:
                end=pivot_index-1


def find_kth_max(k,arr):
    if k<1 or k>len(arr):
        print ('Invalid Input!')
        return -1
    else:
        start=0
        end=len(arr)-1
        while start<=end:
            pivot_index=partition(arr,start,end)
            if pivot_index==(len(arr)-k):
                return arr[pivot_index]
            elif pivot_index<(len(arr)-k):
                start=pivot_index+1
            else:
                end=pivot_index-1

if __name__ == '__main__':
    arr = [10, 2, 7, 4, 9, 6, 3, 8, 1, 5]
    k=9
    print ("K-th MIN : ",find_kth_min(k, arr))
    print ("K-th MAX : ",find_kth_max(k, arr))